[quant][pt2e] Add reference representation for quantized max_pool2d

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105784
* #105783
* #105709
* __->__ #105708

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add reference representation for quantized max_pool2d

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1814,6 +1814,22 @@ def forward(self, x, y):
+        quantizer.set_global(operator_config)
+        m_eager = TestHelperModules.ConvMaxPool2d().eval()
+
+        example_inputs = (torch.randn(1, 2, 2, 2),)
```
Is the replacement failing here because of "burnt" in scalars issue?

jerryzh168 replied:
the replacement succeeds in the test (checked locally), do you mean why the replacement doesn't work before we have `replace_literals_with_placeholders` workaround?

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -93,6 +96,41 @@ def _reference_quantized_add(
+    acc_i8, _ = torch.ops.aten.max_pool2d_with_indices.default(x_i8, kernel_size, stride, padding, dilation, ceil_mode)
+    acc_i32 = acc_i8.to(torch.int32)
+    # TODO: use mul.Scalar, need to change how we handle literal args
+    output_i32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, acc_i32, (x_scale / out_scale))
+    output_i8 = output_i32 - (x_zero_point * x_scale / out_scale + out_zero_point)
+    output_i8 = output_i8.to(torch.int8)
```
I think this should really be simplified with 
```
# these can be dtype specialized by doing zp subtraction in int16 when quantized dtype is int8 and int32 when quantized dtype is int16
x_i32 = x_i8.to(torch.int32) - x_zero_point.to(torch.int32)
acc_i32, _ = torch.ops.aten.max_pool2d_with_indices.default(x_i8, kernel_size, stride, padding, dilation, ceil_mode)
output_i32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, acc_i32, (x_scale / out_scale))
output_i8 = output_i32 - x_zero_point
```
then as a nit requant logic can be factored out.

jerryzh168 replied:
actually this will make input to max_pool2d op int16/int32 right? then we are not doing maxpool2d operation in int8

kimishpatel replied:
I think thats fine because in quantized matmul we do matmul in int16 due to zero point subtraction. But you are also right if backends are actually lowering exactly this representation. That is why the representation is really backend tied. We providing one here should mostly be for conveying numerics.

jerryzh168 replied:
OK, maybe I can use int16 here as well?

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -140,19 +178,29 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
-    (_QUANTIZED_ADD_EXAMPLE_INPUTS, _qdq_quantized_add, _reference_quantized_add),
-    (_QUANTIZE_PER_TENSOR_INT8_EXAMPLE_INPUTS, _quantize_per_tensor_int8, _reference_quantize_per_tensor_int8),
-    (_DEQUANTIZE_PER_TENSOR_INT8_EXAMPLE_INPUTS, _dequantize_per_tensor_int8, _reference_dequantize_per_tensor_int8),
+    (_QUANTIZED_ADD_EXAMPLE_INPUTS, _qdq_quantized_add_relu, _reference_quantized_add_relu, _DONT_REPLACE_LITERAL),
```
```suggestion
    (_QUANTIZED_ADD_RELU_EXAMPLE_INPUTS, _qdq_quantized_add_relu, _reference_quantized_add_relu, _DONT_REPLACE_LITERAL),
```

jerryzh168 replied:
we are using the same thing for both add and add_relu actually.. let me rename this constant

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -140,19 +178,29 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
         replacement = get_aten_graph_module(replacement, example_inputs)  # type: ignore[arg-type, assignment]
         remove_tensor_overload_for_qdq_ops(replacement)  # type: ignore[arg-type]
+        if replace_literals:
+            pattern = replace_literals_with_placeholders(pattern)
```
you said, this is already not working, right?

jerryzh168 replied:
this is working, just a bit hacky

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -237,3 +237,37 @@ def dropout_eval(x):
+from torch.fx.node import map_arg
+from torch.utils._pytree import LeafSpec
+
+def replace_literals_with_placeholders(gm):
```
Something like "This is a special case for replace_pattern where we don't have to manually copy over the literal args. If the above conditions are not satisfied then we will still have to manually copy the args."

jerryzh168 replied:
yeah, the num of literal args and its order of appearance need to match exactly, will add a comment

kimishpatel replied:
we have not concluded on a sustainable solution for this, right?

jerryzh168 replied:
yeah we haven't, current code is short term solution

andrewor14 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -142,19 +180,29 @@ def _reference_dequantize_per_tensor_int8(x_i8, scale, zero_point, quant_min, qu
     # TODO: debug the implementation later when torchdynamo time out issue is resolved
     return ((x_i8.to(torch.float32) - zero_point) * scale).to(dtype=torch.float32)
 
+# boolean flag to indicate if we want to replace literals to be placeholder nodes for pattern and replacement
```
I think we need to explain here as well why we need to do this in the first place

andrewor14 replied:
and maybe add a warning to say this will suddenly fail if the number of literal args or the order of the args are not preserved

jerryzh168 replied:
> and maybe add a warning to say this will suddenly fail if the number of literal args or the order of the args are not preserved

this is probably not possible to do, developers should make sure this is the case I think

jerryzh168 replied:
> I think we need to explain here as well why we need to do this in the first place

this is replaced in future PRs, will add comments later

andrewor14 reviewed (commented):
Graph util changes look good to me. I think we should add some comments to explain why we need to replace literals in the first place and what the risks are. Going to let Kimish stamp this for the other changes.

andrewor14 reviewed (approved):
Stamping for now to unblock

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -95,6 +98,41 @@ def _reference_quantized_add(
+    acc_i32 = acc_i8.to(torch.int32)
+    # TODO: use mul.Scalar, need to change how we handle literal args
+    output_i32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, acc_i32, (x_scale / out_scale))
+    output_i8 = output_i32 - (x_zero_point * x_scale / out_scale + out_zero_point)
```
Also I dont think `(x_zero_point * x_scale / out_scale + out_zero_point)` is correct, right? affine quant is
`fp32 = scale * (int8 - zero_point)` so `int8 = fp32/scale + zero_point)`?

Has this been functionally tested? If not we should add functional tests for each.

jerryzh168 replied:
probably haven't test this one locally, let me check after changing to int16

kimishpatel replied:
I wasnt pointing out the int16 issue here btw. I was pointing out a potential but because in the expression above you are subtracting out_zero_point, but you actually need to add

jerryzh168 replied:
yeah understood, I'll update the representation with int32 maxpool and make sure to check the numerics locally

kimishpatel reviewed (commented):
Left some more comments.

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized max_pool2d"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_maxpool2d

Although right now it is not really testing things since there is some problem with dynamo export

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
